sphinx/testing/_internal/markers.py

```diff
@@ -0,0 +1,364 @@
+    confoverrides: dict[str, Any] | None
+    status: StringIO | None
+    warning: StringIO | None
+    freshenv: bool
+    warningiserror: bool
+    tags: list[str] | None
+    verbosity: int
+    parallel: int
+    keep_going: bool
+    # :class:`sphinx.testing.util.SphinxTestApp` optional arguments
+    builddir: Path | None
+    docutils_conf: str | None
+    # :class:`sphinx.testing.util.SphinxTestApp` extras arguments
+    isolate: Required[Isolation]
+    """The deduced isolation policy."""
+    testroot: Required[str | None]
+    """The deduced testroot ID (possibly None if the default ID is not set)."""
+    testroot_path: Required[str | None]
+    """The absolute path to the testroot directory, if any."""
+    shared_result: Required[str | None]
+    """The optional shared result ID."""
+
+
+class AppParams(NamedTuple):
+    """The processed arguments of :func:`pytest.mark.sphinx`.
+
+    The *args* and *kwargs* values can be directly used as inputs
+    to the :class:`~sphinx.testing.util.SphinxTestApp` constructor.
+    """
+
+    args: list[Any]
+    """The constructor positional arguments, except ``buildername``."""
+    kwargs: SphinxInitKwargs
+    """The constructor keyword arguments, including ``buildername``."""
+
+
+class AppLegacyParams(NamedTuple):
+    args: list[Any]
+    kwargs: dict[str, Any]
+
+
+class TestParams(TypedDict):
+    """A view on the arguments of :func:`pytest.mark.test_params`."""
+
+    shared_result: str | None
+
+
+def _get_sphinx_environ(node: PytestNode, default_builder: str) -> SphinxMarkEnviron:
+    args, kwargs = get_mark_parameters(node, 'sphinx')
+
+    if len(args) > 1:
+        err = 'expecting at most one positional argument'
+        pytest.fail(format_mark_failure('sphinx', err))
+
+    env = cast(SphinxMarkEnviron, kwargs)
+
+    if args:
+        buildername = args.pop()
+        if buildername != env.pop('buildername', buildername):
+            err = '%r has duplicated values' % 'buildername'
+            pytest.fail(format_mark_failure('sphinx', err))
+        env['buildername'] = buildername
+    else:
+        buildername = env.setdefault('buildername', default_builder)
+
+    if not buildername:
+        pytest.fail(format_mark_failure('sphinx', 'missing builder name'))
+
+    check_mark_keywords('sphinx', SphinxMarkEnviron.__annotations__, env, node=node)
+    return env
+
+
+def _get_test_srcdir(
+    srcdir: str | None,
+    testroot: str | None,
+    shared_result: str | None,
+) -> str:
+    """Deduce the sources directory from the given arguments.
+
+    :param srcdir: An optional explicit source directory name.
+    :param testroot: An optional testroot ID to use.
+    :param shared_result: An optional shared result ID.
+    :return: The sources directory name *srcdir* (non-empty string).
+    """
+    check_mark_str_args('sphinx', srcdir=srcdir, testroot=testroot)
+    check_mark_str_args('test_params', shared_result=shared_result)
```

Comment:
I'm not sure that it's worth extracting a utility method for these checks; I think it may be simpler to perform the checks inline.

(this method is used twice here and in part-3, and similarly `check_mark_keywords` is used three times)

Reply:
The reason is mostly because to have always the same message and to avoid possible mistakes.